lib/fallback.ts

```diff
@@ -10,7 +10,74 @@ export type FallbackOptions = {
  * Allows to provide a substitute value when method call fails or rejects.
  * @param options optional fallback options
  */
-export function fallback(valueProvider: (...args: any[]) => any, options?: FallbackOptions);
-export function fallback(value: any, options?: FallbackOptions) {
-  throw new Error('Not implemented.');
+export function fallback(valueProvider: (...args: any[]) => any, options?: FallbackOptions): any;
+export function fallback(value: any, options?: FallbackOptions): any;
+export function fallback(value: any | ((...args: any[]) => any), options?: FallbackOptions): any {
+
+  return function (target: any, propertyKey: any, descriptor: PropertyDescriptor) {
+    const method = descriptor.value;
+
+    descriptor.value = function () {
+      try {
+        const result = method.call(this, arguments);
+        const isPromiseLike = result instanceof Promise;
+
+        if (isPromiseLike) {
+          return fallbackPromise(options, value, result, this);
+        }
+
+        return result;
+
+      } catch (err) {
+        const isFiltered = filterError(err, options);
+
+        if (isFiltered) {
+          return getFallbackValue(value, this);
+        }
+
+        throw new Error(err.message);
+      }
+    };
+
+  };
+}
+
+function fallbackPromise(
+  options: FallbackOptions,
+  value: any,
+  result: any,
+  instance: any,
+): Promise<any> {
+
+  const resolve = (response: any) => {
+    return Promise.resolve(response);
+  };
+
+  const reject = (err: any) => {
+    const isFiltered = filterError(err, options);
+
+    return isFiltered
+      ? Promise.resolve(getFallbackValue(value, instance))
+      : Promise.reject(err);
+  };
+
+  return Promise.resolve(result).then(resolve, reject);
+}
+
+function filterError(error: Error, options: FallbackOptions): boolean {
+  if (!options || !options.errorFilter) {
+    return true;
+  }
+
+  return options.errorFilter(error);
```

Comment:
Would be nice if the error filter function is binded to the instance, same as the fallback function.